File: python/product_distribution_entropy.py
#!/usr/bin/env python
import numpy as np
import argparse

parser = argparse.ArgumentParser(description='Calculate entropy from a 1D distribution.')
parser.add_argument('file', type=str,nargs='?',default="trans_dist.npz",
                           help='file to run on.')
parser.add_argument('nbins', type=int,nargs='?',default=5000,
                           help='Number of bins to generate')
parser.add_argument('-eta', type=float,nargs=1, default=1e-10,
                           help='Small constant to add to dist for log normalization')

args = parser.parse_args()
dist = np.load(args.file)

entropy = 0

for i in range(len(dist[0,:])):
    bins, edges = np.histogram(dist[:,i], bins=args.nbins)

    dx = edges[1] - edges[0]
    P = bins / np.sum(bins) / dx + args.eta
    entropy += -np.trapz(P * np.log(P), dx=dx)
    
#print("Entropy when treating transformed dist as product distribution: ",entropy)
print(entropy)

File: julia/product_distribution_entropy.jl
#!/usr/bin/env julia
using NPZ
using ArgParse

function read_arg()    
    s = ArgParseSettings()    
    @add_arg_table! s begin 
    "file"
        help = "Enter file name of distribution."
        arg_type = String             
    "nbins"
        help = "Enter the number of bins desired for numeric entropy calculation"
        arg_type = Int
    end

    return parse_args(s)
end

#args = read_arg()

#dist = npzread(args["file"])
#nbins = args["nbins"]

function entropy(dist, nbins)
    ent = []    
    
    for i in 1:length(dist[1,:])
        
        x = dist[:,i]

        #find the range of x
        x_max = maximum(x)
        x_min = minimum(x)

        x_range = x_max - x_min

        #turn points into bin numbers
        bin_x = Int.(floor.((x .- x_min)./x_range.*(nbins - 1)).+ 1)

        #set vector for bins
        e = zeros(nbins)

        #place data in respective bins
        for i in 1:length(x)
            e[bin_x[i]] += 1
        end

        #calculate the probability of landing at any index in vector e
        e = e ./ sum(e)
        #add a very small number to avoid error when taking the logarithm
        e = e .+ 10^(-20)

        #calculate the entropy of the system
        dx = x_range/(nbins-1)
        append!(ent, -sum(log.(e).*e)*dx)
    end
    return(sum(ent))
end

#val = entropy(dist, nbins)

print("Entropy assuming is a product distribution: ", val)
